hs/app/reach-cli/Main.hs

```diff
@@ -0,0 +1,352 @@
+upgrade = undefined
+
+
+--------------------------------------------------------------------------------
+cmdUpdate :: Mod CommandFields Cli
+cmdUpdate = command "update" $ info (pure Update) desc where
+  desc = progDesc "Update Reach Docker images"
+
+
+update :: Script ()
+update = undefined
+
+
+--------------------------------------------------------------------------------
+cmdDockerReset :: Mod CommandFields Cli
+cmdDockerReset = command "docker-reset" $ info (pure DockerReset) desc where
+  desc = progDesc "Docker kill and rm all images"
+
+
+dockerReset :: Script ()
+dockerReset = undefined
+
+
+--------------------------------------------------------------------------------
+cmdVersion :: Mod CommandFields Cli
+cmdVersion = command "version" $ info (pure Version) desc where
+  desc = progDesc "Display version"
+
+
+version :: Script ()
+version = cmd "echo" (quote $ "reach " <> reachVersion)
+
+
+--------------------------------------------------------------------------------
+cmdHelp :: Mod CommandFields Cli
+cmdHelp = command "help" $ info (pure Help) desc where
+  desc = progDesc "Show usage"
+
+
+help' :: Script ()
+help' = undefined
+
+
+--------------------------------------------------------------------------------
+cmdHashes :: Mod CommandFields Cli
+cmdHashes = command "hashes" $ info (pure Hashes) desc where
+  desc = progDesc "Display git hashes used to build each Docker image"
+
+
+hashes :: Script ()
+hashes = flip mapM_ reachImages $ \i -> do
+  let t = "reachsh/" <> i <> ":" <> reachVersion
+      s = cmd "docker" "run" "--entrypoint" "/bin/sh" t "-c" (quote "echo $REACH_GIT_HASH")
+  cmd "echo" (i <> ":") (Output s)
+
+
+--------------------------------------------------------------------------------
+cmdWhoami :: Mod CommandFields Cli
+cmdWhoami = command "whoami" $ info (pure Whoami) fullDesc
+
+
+whoami :: Script ()
+whoami = cmd "docker" "info" "--format" "{{.ID}}"
+  |> (stdError, T.unpack "/dev/null")
+
+
+--------------------------------------------------------------------------------
+cmdNumericVersion :: Mod CommandFields Cli
+cmdNumericVersion = command "numeric-version" $ info (pure NumericVersion) fullDesc
+
+
+numericVersion :: Script ()
+numericVersion = undefined
+
+
+--------------------------------------------------------------------------------
+cmdReactDown :: Mod CommandFields Cli
+cmdReactDown = command "react-down" $ info (pure ReactDown) fullDesc
+
+
+reactDown :: Script ()
+reactDown = undefined
+
+
+--------------------------------------------------------------------------------
+cmdRpcServerDown :: Mod CommandFields Cli
+cmdRpcServerDown = command "rpc-server-down" $ info (pure RpcServerDown) fullDesc
+
+
+rpcServerDown :: Script ()
+rpcServerDown = undefined
+
+
+--------------------------------------------------------------------------------
+cmdUnscaffold :: Mod CommandFields Cli
+cmdUnscaffold = command "unscaffold" $ info (pure Unscaffold) fullDesc
+
+
+unscaffold :: Script ()
```

Comment:
I think you should use a type to encapsulate what a command is expected to be

Reply:
I've created `type Subcommand = ParserInfo (Script ())` in response to the other thread but I don't think that's really what you're asking for here...

Can you foresee a case where we'd need to pass something other than `()` up to the `optparse` interpreter?

Reply:
How does the `()` go up to the interpreter? I thought the `()` is the return from the shell script

Reply:
I basically just mean this line: `join . fmap sh $ customExecParser (prefs showHelpOnError) cmds`.

The scripts don't really return anything meaningful back to the Haskell code.